File: Invert-Binary-Tree/Invert-Binary-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var invertTree = function (root) {
  /* Recursive
   if (root) {
   [root.left, root.right] = [invertTree(root.right), invertTree(root.left)];
   }
   return root;
   */

  if (root === null) {
    return null;
  }
  const queue = [];
  queue.push(root);
  while (queue.length) {
    const temp = queue.shift();
    const left = temp.left;
    temp.left = temp.right;
    temp.right = left;
    if (temp.left !== null) {
      queue.push(temp.left);
    }
    if (temp.right !== null) {
      queue.push(temp.right);
    }
  }
  return root;
};

File: solutions/Invert-Binary-Tree/Invert-Binary-Tree.cpp
// 2019-07-28
// O(n)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
        if(!root)
        {
            return root;
        }
        auto tmp = root->left;
        root->left = invertTree(root->right);
        root->right = invertTree(tmp);
        return root;
    }
};